sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -960,8 +960,27 @@ object TableFunctionRegistry {
     (name, (info, (expressions: Seq[Expression]) => builder(expressions)))
   }
 
+  def generator[T <: Generator : ClassTag](name: String, outer: Boolean = false)
+      : (String, (ExpressionInfo, TableFunctionBuilder)) = {
+    val (info, builder) = FunctionRegistryBase.build[T](name, since = None)
+    val newBuilder = (expressions: Seq[Expression]) => {
+      val generator = builder(expressions)
+      assert(generator.isInstanceOf[Generator])
+      Generate(
+        generator,
+        unrequiredChildIndex = Nil,
+        outer = outer,
+        qualifier = None,
+        generatorOutput = Nil,
```

Comment:
is it guaranteed that we will assign the `generatorOutput` later?

Reply:
Yes. Generate is only resolved when its output length equals to the generator.elementSchema length.
https://github.com/apache/spark/blob/6e537f30ac2b19617c6cc11dd4f7f92946988f7b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala#L267-L272
The Generate created here will be resolved by `ResolveGenerate`.